pkg/pyramid/tier_fs.go

```diff
@@ -114,7 +115,12 @@ func (tfs *TierFS) removeFromLocal(rPath params.RelativePath, filesize int64) {
 	// This will be called by the cache eviction mechanism during entry insert.
 	// We don't want to wait while the file is being removed from the local disk.
 	evictionHistograms.WithLabelValues(tfs.fsName).Observe(float64(filesize))
-	go tfs.removeFromLocalInternal(rPath)
+	// Decrement the reference count on the file. When the ref count reaches -1, The file will be deleted
+	tfs.fileTracker.dec(rPath, func(ref int) {
+		if ref == -1 {
+			tfs.removeFromLocalInternal(rPath)
```

Comment:
What happens if `ref` returns as `-1` and before the file is removed, it is acquired again, by another thread.
I think it will still be removed, as the mutex protects the ref-count, but nothing prevents ref increase, after the `-1` is returned and `removeFromLocalInternal` is fired

Reply:
This cannot happen because removeFromLocalInternal happens in the scope of dec where we have a mutex lock on the map.
What will happen is that on dec the entry will be deleted from the map, and the file deleted from storage. Only after that the lock will be released - the inc will happen and a new entry in the map will be created for the file